### PR TITLE
Ntjoin multi-ref fix

### PR DESCRIPTION
**Pull request by ycecilia**
I've added a new Python script for a couple of tasks: first, it creates separate reference FASTA files, and second, it generates a config csv file so that ntjoin can access these FASTA files all at once. 

Also, I've included an extra gap check (in mtgrasp.smk) after running ntjoin in case it fails and we don't need to use Sealer.

### CODE REVIEW

File: README.md [hunk 1]
```diff
@@ -40,8 +40,25 @@ git clone https://github.com/bcgsc/mtGrasp.git
 * ntCard
 
 ---
+### Installation Instructions for Dependencies
+#### Step 1: 
 
-#### Special Installation Instructions for MITOS
+Recommended (Faster download):
```
Comment (lcoombe) [inline README.md:46]:
```suggestion
Recommended (Faster):
```
Like this new section a lot!

File: README.md [hunk 2]
```diff
@@ -40,8 +40,25 @@ git clone https://github.com/bcgsc/mtGrasp.git
...
+```
+conda create -n mtgrasp python=3.10 mamba
+conda activate mtgrasp
+mamba install -c conda-forge -c bioconda snakemake 'blast>=2.10.0' biopython seqtk abyss ntjoin bwa samtools pilon ntcard
+```
+
+Alternative (Slower download):
```
Comment (lcoombe) [inline README.md:53]:
```suggestion
Alternative (Slower):
```

File: README.md [hunk 3]
```diff
@@ -193,28 +210,6 @@ If you are not interested in the standardized mitogenome sequence(s) or MITOS an
 
 
 
-### Summarize mtGrasp results
```
Comment (lcoombe) [inline README.md:196]: I thought that the script just needed some fixes? Why are we deleting it altogether?

File: create_references_for_ntjoin.py [hunk 4]
```diff
@@ -0,0 +1,27 @@
...
+# Usage: split_fasta.py <fasta file> <output directory> <config file>
+
+import sys
+
+input = sys.argv[1]
+output = sys.argv[2]
+config = sys.argv[3]
+ref_list = []
+with open(input, 'r') as f:
+    lines = f.readlines()
```
Comment (lcoombe) [inline create_references_for_ntjoin.py:13]: It's good practice to read through files line-by-line, versus reading everything into memory, which isn't really necessary (and can inflate the memory usage). You can just see if the line is a header, and store it in a variable if so, ready to write the sequence once you get to the next line.
```suggestion
    for line in f:
        line = line.strip()
        // do the logic
```

File: create_references_for_ntjoin.py [hunk 5]
```diff
@@ -0,0 +1,27 @@
...
+    for i in range(len(lines)):
+        if lines[i].startswith('>'):
+            header = lines[i].strip()
+            filename = header.split(' ')[0].replace('>', '') + '.fasta'
+            sequence = lines[i+1].strip()
+            with open(output + '/' + filename, 'w') as g:
+                g.write(header + '\n')
+                g.write(sequence + '\n')
+            ref_list.append(f'{output}/{filename},2')
```
Comment (lcoombe) [inline create_references_for_ntjoin.py:22]: I'd make this a variable at the top of the file - it'll be easier in the future to remember what this number is for our future selves, and a tad more straightforward for someone to change if needed

File: create_references_for_ntjoin.py [hunk 6]
```diff
@@ -0,0 +1,27 @@
...
+input = sys.argv[1]
+output = sys.argv[2]
+config = sys.argv[3]
+ref_list = []
+with open(input, 'r') as f:
+    lines = f.readlines()
+    for i in range(len(lines)):
+        if lines[i].startswith('>'):
+            header = lines[i].strip()
+            filename = header.split(' ')[0].replace('>', '') + '.fasta'
```
Comment (lcoombe) [inline create_references_for_ntjoin.py:17]: Just use `strip(">")`

File: mtgrasp.smk [hunk 7]
```diff
@@ -202,14 +202,18 @@ rule extract_seq:
...
          current_dir + "{library}/benchmark/k{k}_kc{kc}.extract_seq.benchmark.txt"
      params:
-         ref_fasta=config["ref_path"]
+         ref_fasta=config["ref_path"],
+         ref_outdir=current_dir + "{library}/blast/mito_refs",
+         ref_config = current_dir + "{library}/blast/k{k}_kc{kc}-ntjoin_ref_config.csv"
      shell:
          "seqtk subseq {input.assemblies} {input.query} > {output.query_out} ; "
-         "seqtk subseq {params.ref_fasta} {input.ref} > {output.ref_out}"
+         "seqtk subseq {params.ref_fasta} {input.ref} > {output.ref_out} &&"
```
Comment (lcoombe) [inline mtgrasp.smk:214]: `&&` shouldn't be needed? Could just use `;`
Ideally, these should be separate rules - this is now doing sort of separate things (extracting sequences and creating the config file for ntJoin). But if it'll take too long to change that, OK to leave as-is

**Review by lcoombe — COMMENTED** [review]
Looks good Cecilia! Just a few comments/suggestions :)

File: README.md (hunk 3 above)
Comment (ycecilia) [reply README.md:196]: I will add this section once i have the summary.py is ready

File: create_references_for_ntjoin.py (hunk 6 above)
Comment (ycecilia) [reply create_references_for_ntjoin.py:19]: The headers usually have space between texts, so If we were to use `strip(">")`, the filename would end up being something like `NC.1234.1 species name complete mitochondrion.fasta` which is not ideal. `filename = header.split('  ')[0].replace('>', '') + '.fasta'`, on the other hand, creates `NC.1234.1.fasta`.

Comment (lcoombe) [reply create_references_for_ntjoin.py:19]: But you are already splitting by a space and selecting the first in the list?

File: create_references_for_ntjoin.py (hunk 5 above)
Comment (ycecilia) [reply create_references_for_ntjoin.py:24]: Did you mean creating a `reference_weight` variable? Do you think there will be situations where references need to be assigned different weight values? Or do you think it is not necessary?

File: create_references_for_ntjoin.py (hunk 6 above)
Comment (ycecilia) [reply create_references_for_ntjoin.py:19]: what about` filename = header.strip('>').split(' ')[0] + '.fasta'`

Comment (lcoombe) [reply create_references_for_ntjoin.py:19]: Yup, either that or `filename = header.split(' ')[0].strip('>') + '.fasta'` (that's what I meant initially)

File: create_references_for_ntjoin.py (hunk 5 above)
Comment (lcoombe) [reply create_references_for_ntjoin.py:24]: Yes, that's what I mean! This is a 'magic number' as-is (like when a number seems to come out of nowhere), so giving it a variable name instead makes the code more readable

File: create_references_for_ntjoin.py (hunk 6 above)
Comment (ycecilia) [reply create_references_for_ntjoin.py]: Done

File: README.md (hunk 3 above)
Comment (ycecilia) [reply README.md:196]: Added summarize.py.

**Pull request merged**